### PR TITLE
test case(authentication): get user token

### PR DESCRIPTION
**Pull request by bhaskarhc**
Test Kubera  local authentication using user credentials 

### How to run the test : 
 -  Credentials and Kubera host details need to set as environment variables
```console
KUBERA_HOST="<Kubera_IP>" \
  KUBERA_USER="<kubera_user_name>" \
  KUBERA_PASS="<kubera_user_password>" \
  go test ./... 
```

#### Folder Structure
```
|_ kubera-api-testing/
    |__ kubera/
    |   |__ connection/
    |   |    |__ connection.go
    |   |__ authentication/
    |        |__ authentication_test.go
    |        |__ authentication.go
    |__ go.mod
    |__ go.sum
    |__ makefile
```

#### Scenarios
```json
{
		"No host": {
			Username: UserName,
			Password: Password,
			IsError:  true,
		},
		"No username": {
			Hostname: HostName,
			Password: Password,
			IsError:  true,
		},
		"No password": {
			Hostname: HostName,
			Username: UserName,
			IsError:  true,
		},
		"With host, username & password": {
			Hostname: HostName,
			Username: UserName,
			Password: Password,
		},
		"With host, username & invalid password": {
			Hostname: HostName,
			Username: UserName,
			Password: "invalid",
			IsError:  true,
		},
		"With invalid host, username & password": {
			Hostname: "invalid",
			Username: UserName,
			Password: Password,
			IsError:  true,
		},
		"With host, invalid username & password": {
			Hostname: HostName,
			Username: "invalid",
			Password: Password,
			IsError:  true,
		},
	}
```

#### Output
```console
bhaskarjb@Bhaskars-MBP:~/Desktop/mayadata-io/kubera-api-testing|get-user-data 
⇒  KUBERA_HOST="http://15.207.111.248/" KUBERA_USER="test1@e2e.io" KUBERA_PASS="Test@e2e.io" go test ./...    
ok  	github.com/mayadata-io/kubera-api-testing/kubera/authentication	(cached)
?   	github.com/mayadata-io/kubera-api-testing/kubera/connection	[no test files]
```

### CODE REVIEW

File: testing_api/rest.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
+package rest
```
Comment (AmitKumarDas) [inline testing_api/rest.go]: what is the reason for the name testing_api?

File: testing_api/rest.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+import (
+	"fmt"
+
+	"github.com/go-resty/resty/v2"
+)
+
+// SuccessRequest for ,if request success then to get token
+type SuccessRequest struct {
```
Comment (AmitKumarDas) [inline testing_api/rest.go]: This does not explain. I was not able to understand the comment

File: testing_api/rest.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	AccountID string `json:"accountId"`
+}
+
+// FailedRequest fail return error
+type FailedRequest struct {
+	Msg string `json:"message"`
+}
+
+// UserData contains the details of credentials
+type UserData struct {
```
Comment (AmitKumarDas) [inline testing_api/rest.go]: Are we modelling some specific things?
Is it KuberaUser or DirectorUser or something else?
Let us try to put specific names for structures, variables & so on

File: testing_api/rest.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	UserName string
+	Password string
+	Host     string
+}
+
+// PostResponse for defining the response of postmethod
+type PostResponse struct {
```
Comment (AmitKumarDas) [inline testing_api/rest.go]: Can we rename this to Response only?
Do we feel the need to have GetResponse, DeleteResponse, etc?

File: testing_api/rest.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	Resp       *resty.Response
+	Token      string
```
Comment (AmitKumarDas) [inline testing_api/rest.go]: Please explain each of the fields

File: testing_api/Makefile [hunk 6]
```diff
@@ -0,0 +1,7 @@
+.PHONY: test install
```
Comment (AmitKumarDas) [inline testing_api/Makefile]: move out to root folder

File: testing_api/rest.go (hunk 1 above)
Comment (AmitKumarDas) [inline testing_api/rest.go]: scenarios/authentication should be the folder structure

File: testing_api/rest.go (hunk 2 above)
Comment (AmitKumarDas) [reply testing_api/rest.go]:
```go
  type KuberaToken struct {
    JWT
    AccountID
  }
  
  type Tokenisable struct {
    URL string
  }
  
  func NewTokensier() *Tokenisable {
    return &Tokenisable{}
  }
  
  func (a *Tokenisable) GetKuberaToken() KuberaToken {
   // ...
  }
  
  --- _test.go
  
  t := NewTokensier()
  token := t.GetKuberaToken()
```

File: testing_api/rest.go [hunk 7]
```diff
@@ -1,24 +1,32 @@
...
+	"os"
 
 	"github.com/go-resty/resty/v2"
 )
 
-// SuccessRequest for ,if request success then to get token
-type SuccessRequest struct {
-	Token     string `json:"jwt"`
+const (
+	kuberahost = "KHOST"
```
Comment (AmitKumarDas) [inline testing_api/rest.go]: Let us make use of `KUBERA_HOST` as env

File: testing_api/rest.go [hunk 8]
```diff
@@ -1,24 +1,32 @@
...
+	kuberauser = "KUSER"
```
Comment (AmitKumarDas) [inline testing_api/rest.go]: same comment

File: testing_api/rest.go [hunk 9]
```diff
@@ -35,35 +43,53 @@ type PostResponse struct {
 	// Body    []byte
 }
 
-// type HttpPost interface {
-// 	getToken() PostResponse
-// }
+//NewTokensier return the credentials of kubera user ,pass and host
+func NewTokensier() (*Tokenisable, error) {
+	host, ok := os.LookupEnv(kuberahost)
```
Comment (AmitKumarDas) [inline testing_api/rest.go]: try making use of flags & if flag is not set then ENV

File: testing_api/rest.go [hunk 10]
```diff
@@ -35,35 +43,53 @@ type PostResponse struct {
...
+	fmt.Println(host, userName, userPass)
+	return &Tokenisable{
+		Host:     host,
+		UserName: userName,
+		Password: userPass,
+	}, nil
+}
 
-func (u UserData) getToken() *PostResponse {
+func (a *Tokenisable) GetKuberaToken() *PostResponse {
```
Comment (AmitKumarDas) [inline testing_api/rest.go]: This should return KuberaToken & err

File: testing_api/rest.go [hunk 11]
```diff
@@ -1,24 +1,32 @@
...
-type FailedRequest struct {
-	Msg string `json:"message"`
+type KuberaError struct {
+	Message string `json:"message"`
 }
 
-// UserData contains the details of credentials
-type UserData struct {
+// Tokenisable contains the details of credentials
+type Tokenisable struct {
```
Comment (AmitKumarDas) [inline testing_api/rest.go]:
```go
  type TokenFetching struct {}
  func NewTokenFetcher() *TokenFetching {}
  func (t *TokenFetching) FetchKuberaToken() (KuberaToken, error) {}
```

File: testing_api/rest.go [hunk 12]
```diff
@@ -75,7 +101,7 @@ type GetReqResponse struct {
 
 // getRequest func is to make GET method requests
 func (p PostResponse) getRequest(api string) *GetReqResponse {
```
Comment (AmitKumarDas) [inline testing_api/rest.go]: Can you rename PostResponse?
Make it readable

Comment (AmitKumarDas) [inline testing_api/rest.go]:
```go
type Request struct {}
type Response struct {}
type Invocable struct {}
func NewInvoker() *Invocable {}
func (i *Invocable) Invoke() (Response, error) {} 
```

File: scenarios/authentication/rest.go [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+	}
+	userName, ok := os.LookupEnv(kuberauser)
+	if !ok {
+		return nil, errors.New("KUSER environment variable required")
+	}
+	userPass, ok := os.LookupEnv(kuberapass)
+	if !ok {
+		return nil, errors.New("KPASS environment variable required")
+	}
+	fmt.Println(host, userName, userPass)
```
Comment (bhaskarhc) [inline scenarios/authentication/rest.go]: use log instead

File: scenarios/authentication/rest.go [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+	return &TokenFetching{
+		Host:     host,
+		UserName: userName,
+		Password: userPass,
+	}, nil
+}
+
+func (a *TokenFetching) FetchKuberaToken() (*Response, error) {
+	client := resty.New()
+	url := a.Host + "v3/token"
```
Comment (bhaskarhc) [inline scenarios/authentication/rest.go]: rm hardcode

File: scenarios/authentication/rest_test.go [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+	"testing"
+)
+
+func TestGetUserDetails(t *testing.T) {
+
+	cred, err := NewTokensier()
+	if err != nil {
+		t.Errorf("Error : %s", err)
+	}
+	tkn, err := cred.FetchKuberaToken()
```
Comment (bhaskarhc) [inline scenarios/authentication/rest_test.go]: cred.FetchKuberaToken().getRequest(cred.Host + "v3/account/")
// can i use like this ??

File: Makefile [hunk 16]
```diff
@@ -0,0 +1,10 @@
+.PHONY: test install run
+
+run:
```
Comment (AmitKumarDas) [inline Makefile]: can we remove this target completely

File: scenarios/authentication/rest.go [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+	"errors"
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/go-resty/resty/v2"
+)
+
+const (
+	kuberahost = "KUBERA_HOST"
```
Comment (AmitKumarDas) [inline scenarios/authentication/rest.go]: explain in comments on what they are
Make them CamelCased

File: scenarios/authentication/rest.go [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+	kuberauser = "KUBERA_USER"
+	kuberapass = "KUBERA_PASS"
+)
+
+type KuberaToken struct {
```
Comment (AmitKumarDas) [inline scenarios/authentication/rest.go]: Let rename this to Token since everything is related to Kubera

File: scenarios/authentication/rest.go [hunk 19]
```diff
@@ -0,0 +1,125 @@
+package rest
```
Comment (AmitKumarDas) [inline scenarios/authentication/rest.go]: lets rename `scenarios` folder to `kubera`

File: scenarios/authentication/rest.go [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+	kuberauser = "KUBERA_USER"
+	kuberapass = "KUBERA_PASS"
+)
+
+type KuberaToken struct {
+	JWT       string `json:"jwt"`
+	AccountID string `json:"accountId"`
+}
+
+type KuberaError struct {
```
Comment (AmitKumarDas) [inline scenarios/authentication/rest.go]: Let us rename to `Error`

File: scenarios/authentication/rest.go [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+	Message string `json:"message"`
+}
+
+// TokenFetching contains the details of credentials
+type TokenFetching struct {
+	UserName string
```
Comment (AmitKumarDas) [inline scenarios/authentication/rest.go]:
```go
type Connection struct {
  Host string
  Username string
  Password string
}

type TokenFetching struct {
  Connection
}
```

File: scenarios/authentication/rest.go [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+	Password string
+	Host     string
+}
+
+//RequestAccess have a credentials to access Kubera api
+type RequestAccess struct {
```
Comment (AmitKumarDas) [inline scenarios/authentication/rest.go]: Remove this.

File: scenarios/authentication/rest.go [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+	Token     string
+	AccountID string
+}
+
+// Response for defining the response of postmethod
+type Response struct {
```
Comment (AmitKumarDas) [inline scenarios/authentication/rest.go]: Remove this

File: scenarios/authentication/rest.go [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+	Resp       *resty.Response
+	Token      string
+	Message    string
+	Error      error
+	StatusCode int
+	AccountID  string
+}
+
+//NewTokensier return the credentials of kubera user ,pass and host
+func NewTokensier() (*TokenFetching, error) {
```
Comment (AmitKumarDas) [inline scenarios/authentication/rest.go]:
```go
type TokenFetchingConfig struct {
  Connection
}

type TokenFetching struct {
  Connection
}

func NewTokenFetcher(config TokenFetchingConfig) (*TokenFetching, error) {
  // ...
}
```

File: scenarios/authentication/rest.go [hunk 25]
```diff
@@ -0,0 +1,125 @@
...
+	host, ok := os.LookupEnv(kuberahost)
```
Comment (AmitKumarDas) [inline scenarios/authentication/rest.go]: Lookups will not be done here.
Let `func init() {}` do this

File: scenarios/authentication/rest.go [hunk 26]
```diff
@@ -0,0 +1,125 @@
...
+	}
+	fmt.Println(host, userName, userPass)
+	return &TokenFetching{
+		Host:     host,
+		UserName: userName,
+		Password: userPass,
+	}, nil
+}
+
+func (a *TokenFetching) FetchKuberaToken() (*Response, error) {
```
Comment (AmitKumarDas) [inline scenarios/authentication/rest.go]:
```go
func (f *TokenFetcher) Fetch() (Token, error) {
  // ...
}
```

File: scenarios/authentication/rest.go [hunk 27]
```diff
@@ -0,0 +1,125 @@
...
+		SetBody(bodyData).
+		SetError(&kErr).
+		SetResult(&token).
+		Post(url)
+
+	if resp.StatusCode() != 201 {
+		return nil, errors.New(kErr.Message)
+	}
+
+	return &Response{
```
Comment (AmitKumarDas) [inline scenarios/authentication/rest.go]: This should return `Token` & error

File: scenarios/authentication/rest.go [hunk 28]
```diff
@@ -0,0 +1,125 @@
...
+		"authProvider": "localAuthConfig",
+	}
+	resp, err := client.R().
+		SetHeader("Content-Type", "application/json").
+		SetBody(bodyData).
+		SetError(&kErr).
+		SetResult(&token).
+		Post(url)
+
+	if resp.StatusCode() != 201 {
```
Comment (AmitKumarDas) [inline scenarios/authentication/rest.go]:
```go
type KuberaError struct {
  Err error
  Message string
  StatusCode int
}

if err != nil || kErr.Message != "" || response.StatusCode() != 200 {
  return nil, &KuberaError{
    Err: err,
    Message: kErr.Message,
    StatusCode: response.StatusCode(),
  }
}
```

File: scenarios/authentication/rest.go [hunk 29]
```diff
@@ -0,0 +1,125 @@
...
+		UserName: userName,
+		Password: userPass,
+	}, nil
+}
+
+func (a *TokenFetching) FetchKuberaToken() (*Response, error) {
+	client := resty.New()
+	url := a.Host + "v3/token"
+	token := KuberaToken{}
+	kErr := KuberaError{}
```
Comment (AmitKumarDas) [inline scenarios/authentication/rest.go]:
```go
errmsg := ""
```

File: scenarios/authentication/rest.go [hunk 30]
```diff
@@ -0,0 +1,125 @@
...
+	bodyData := map[string]interface{}{
+		"code":         fmt.Sprintf("%s:%s", a.UserName, a.Password),
+		"authProvider": "localAuthConfig",
+	}
+	resp, err := client.R().
+		SetHeader("Content-Type", "application/json").
+		SetBody(bodyData).
+		SetError(&kErr).
```
Comment (AmitKumarDas) [inline scenarios/authentication/rest.go]:
```go
SetError(&errmsg)
```

Comment (AmitKumarDas) [reply scenarios/authentication/rest.go]: What is the argument's datatype required by `SetError` method?

Comment (AmitKumarDas) [reply scenarios/authentication/rest.go]: Understand the difference between SetError & err response

File: kubera/authentication/rest.go [hunk 31]
```diff
@@ -0,0 +1,99 @@
...
+)
+
+const (
+	kuberaHost = "KUBERA_HOST" //KUBERA_HOST is Env variable contains value of Kubera's Domain/IP
+	kuberaUser = "KUBERA_USER" // Kubera Username
+	kuberaPass = "KUBERA_PASS" //Kubera password
+)
+
+type Token struct {
+	Token     string `json:"jwt"`
```
Comment (AmitKumarDas) [inline kubera/authentication/rest.go]: JWT

File: kubera/authentication/rest.go [hunk 32]
```diff
@@ -0,0 +1,99 @@
...
+	Connection
+}
+
+type TokenFetching struct {
+	Connection
+}
+
+//NewTokenFetcher return the credentials of kubera user ,pass and host
+func NewTokenFetcher() (*TokenFetching, error) {
+	host, ok := os.LookupEnv(kuberaHost)
```
Comment (AmitKumarDas) [inline kubera/authentication/rest.go]: All the env lookups should be populated via `init`

File: kubera/authentication/rest.go [hunk 33]
```diff
@@ -0,0 +1,99 @@
...
+	token := Token{}
+	errMsg := Error{}
+	bodyData := map[string]interface{}{
+		"code":         fmt.Sprintf("%s:%s", a.UserName, a.Password),
+		"authProvider": "localAuthConfig",
+	}
+	resp, err := client.R().
+		SetHeader("Content-Type", "application/json").
+		SetBody(bodyData).
+		SetError(&errMsg). //SetError method is to register the global or client common `Error` object into Resty. It is used for automatic unmarshalling if response status code is greater than 399 and content type either JSON or XML.
```
Comment (AmitKumarDas) [inline kubera/authentication/rest.go]: comments on the top & not on the side
all lines of code & comment should be within 80 chars

File: kubera/authentication/signup.txt [hunk 34]
```diff
@@ -0,0 +1,76 @@
+// package rest
+
```
Comment (amitbhatt818) [inline kubera/authentication/signup.txt]: Can you remove this file Signup.txt if it is not required .

**Review by amitbhatt818 — CHANGES_REQUESTED** [review]

File: kubera/authentication/rest.go [hunk 35]
```diff
@@ -0,0 +1,99 @@
...
+
+	if err != nil || errMsg.Message != "" || resp.StatusCode() != 201 {
+		return nil, &KuberaError{
+			Err:        err,
+			Message:    errMsg.Message,
+			StatusCode: resp.StatusCode(),
+		}
+	}
+
+	return &Token{
```
Comment (AmitKumarDas) [inline kubera/authentication/rest.go]: return token, err

File: kubera/authentication/rest.go [hunk 36]
```diff
@@ -0,0 +1,99 @@
...
+		Connection{
+			Host:     host,
+			UserName: userName,
+			Password: userPass,
+		},
+	}, nil
+}
+
+//Fetch token
+func (a *Connection) Fetch() (*Token, *KuberaError) {
```
Comment (AmitKumarDas) [inline kubera/authentication/rest.go]: return Token, KuberaError

File: kubera/authentication/rest_test.go [hunk 37]
```diff
@@ -0,0 +1,28 @@
...
+
+import (
+	"log"
+	"testing"
+)
+
+var config Connection
+
+func init() {
+	fetchConnection, err := NewTokenFetcher()
```
Comment (AmitKumarDas) [inline kubera/authentication/rest_test.go]: why fetchConnection?

File: kubera/authentication/rest_test.go [hunk 38]
```diff
@@ -0,0 +1,28 @@
...
+	if err != nil {
+		log.Fatalln(err)
+		return
+	}
+	log.Printf("\n UserName : %s ,\n Password : %s, \n Host : %s ", fetchConnection.Host, fetchConnection.UserName, fetchConnection.Host)
+	config = fetchConnection.Connection
```
Comment (AmitKumarDas) [inline kubera/authentication/rest_test.go]: Don't use Token to fetch connection.
Use init or sync.Once to populate connection details from ENV

File: kubera/authentication/rest_test.go [hunk 39]
```diff
@@ -0,0 +1,28 @@
+package rest
```
Comment (rohan2794) [inline kubera/authentication/rest_test.go]: In `rest.go` and `rest_test.go ` why we are using same `package rest`, any specific reason

File: kubera/authentication/rest_test.go (hunk 38 above)
Comment (AmitKumarDas) [reply kubera/authentication/rest_test.go]: Alternatively you may use sync.Once function to populate connection.
Remember we want to populate / fillup the connection . In this case we want to populate via ENV.


In following example I have made use of sync.Once
```go
type Connection struct {
  Hostname string
  Username string
  Password string
}

var connection Connection

func GetConnection() Connection {
  if connection.Hostname != "" {
    return connection
  }
  sync.Once(func(){
    connection = Connection{
      Hostname: env.Get(KUBERA_HOST_NAME)
      Username: env.Get(KUBERA_USER_NAME)
      Password: env.Get(KUBERA_PASSWORD)
    }
  })
  return connection
}
```

File: kubera/authentication/rest_test.go (hunk 39 above)
Comment (bhaskarhc) [reply kubera/authentication/rest_test.go]: @rohan2794 pls goto the [Package names](https://blog.golang.org/package-names) Page , u ll get a clear view on Packages

File: kubera/authentication/rest.go (hunk 36 above)
Comment (bhaskarhc) [reply kubera/authentication/rest.go]: if it is not a pointer, i think can't return  `nil` .. ! 😕  
pls give me suggestion how to do @AmitKumarDas .

File: kubera/authentication/rest_test.go (hunk 39 above)
Comment (rohan2794) [reply kubera/authentication/rest_test.go]: In docs i find below statements in package naming convention in Go
`Another convention is that the package name is the base name of its source directory; the package in src/encoding/base64 is imported as "encoding/base64" but has name base64, not encoding_base64 and not encodingBase64.`

File: kubera/authentication/rest.go (hunk 36 above)
Comment (AmitKumarDas) [reply kubera/authentication/rest.go]: return Token{}, KuberaError

**Review by harshshekhar15 — APPROVED** [review]
/lgtm

**Review by amitbhatt818 — APPROVED** [review]
/lgtm

File: kubera/authentication/authentication_test.go [hunk 40]
```diff
@@ -0,0 +1,20 @@
+package authentication
+
+import (
+	"testing"
+
+	conn "github.com/mayadata-io/kubera-api-testing/kubera/connection"
+)
+
+func TestFetchToken(t *testing.T) {
+
```
Comment (AmitKumarDas) [inline kubera/authentication/authentication_test.go]:
```go
func TestFetchToken(t *testing.T) {
 connection := conn.Get()
 var tests = map[string]struct{
   Hostname string
   Username string
   Password string  
   IsError bool
 }{
  "No host": {
     Username: connection.Username,
     Password: connection.Password,
     IsError: true,
   },
  "No username": {
     Hostname: connection.Hostname,
     Password: connection.Password,
     IsError: true,
  },
  "No password": {
     Hostname: connection.Hostname,
     Username: connection.Username,
     IsError: true,
  },
  "With host, username & password": {
     Hostname: connection.Hostname,
     Username: connection.Username,
     Password: connection.Password,
  },
 }
 for name, mock := range tests {
   name := name
   mock := mock
   t.Run(name, func(t *testing.T){
       fetcher := NewTokenFetcher(TokenFetchingConfig{
         Connection: Connection{
            Hostname: mock.Hostname,
            Username: mock.Username,
            Password: mock.Password,
         },
	})
	tkn, err := fetcher.Fetch()
	if mock.IsError && err.Err == nil {
          t.Fatalf("Expected error got none: %s: \nFetcher=%s: \nToken=%s", fetcher, tkn)
	}
	if !mock.IsError && err.Err != nil {
          t.Fatalf("Expected none error got %s: \nFetcher=%s", err, fetcher)
	}
   })
 }
}
```

Comment (rohan2794) [reply kubera/authentication/authentication_test.go]: @AmitKumarDas what is `mock` here ? Where to use and how ?

**Review by rohan2794 — APPROVED** [review]
LGTM

File: Makefile [hunk 41]
```diff
@@ -0,0 +1,7 @@
+.PHONY: test install
+
+install:
+	go get -t -v ./...
+
+test: install
+	go test -race -cover -v ./...
```
Comment (AmitKumarDas) [inline Makefile]: remove -v

Comment (bhaskarhc) [reply Makefile]: done 👍

File: kubera/authentication/authentication_test.go (hunk 40 above)
Comment (bhaskarhc) [reply kubera/authentication/authentication_test.go]: @rohan2794  
In GoLang range loop syntax will be , 
```go
for index, item := range arrayItems {
}
```
while using MAPS iteration ,
 **The first value returned in map is key and the second value is the value of the key-value pair in map.**
@AmitKumarDas correct it if im wrong .

File: kubera/authentication/rest_test.go (hunk 39 above)
Comment (bhaskarhc) [reply kubera/authentication/rest_test.go]: `"encoding/base64"`  is a standard golang library package . like this package importing  im also not clear .  

please checkout [Blog_on_packages](https://medium.com/rungo/everything-you-need-to-know-about-packages-in-go-b8bac62b74cc) for better understanding .

File: scenarios/authentication/rest.go (hunk 30 above)
Comment (bhaskarhc) [reply scenarios/authentication/rest.go]: _**SetError**_ method is to register the global or client common `Error` object into Resty. It is used for automatic unmarshalling if response status code is greater than 399 and content type either JSON or XML. Can be pointer or non-pointer.

_**SetResult**_ method is to register the response `Result` object for automatic unmarshalling for the request, if response status code is between 200 and 299 and content type either JSON or XML.

File: testing_api/rest.go (hunk 1 above)
Comment (bhaskarhc) [reply testing_api/rest.go]: Updated to New directory structure ,
```
|_ kubera-api-testing/
    |__kubera
      |__ connection/
      |    |__ connection.go
      |__ authentication/
           |__ authentication_test.go
           |__ authentication.go
```

File: Makefile [hunk 42]
```diff
@@ -0,0 +1,7 @@
+.PHONY: test install
+
+install:
+	go get -t -v ./...
```
Comment (prabhatkumarthakur) [inline Makefile:4]: we need to remove the verbose mode

File: kubera/authentication/authentication.go [hunk 43]
```diff
@@ -0,0 +1,111 @@
...
+		" ",
+		".",
+	)
+	if err != nil {
+		panic(err)
+	}
+	return string(raw)
+}
+
+type APIError struct {
```
Comment (prabhatkumarthakur) [inline kubera/authentication/authentication.go:28]: can we have a comment that what the below structure will be doing?

**Review by amitbhatt818 — APPROVED** [review]
/lgtm

File: Makefile (hunk 42 above)
Comment (bhaskarhc) [reply Makefile:4]: will add these changes in next PR

File: kubera/authentication/authentication.go (hunk 43 above)
Comment (bhaskarhc) [reply kubera/authentication/authentication.go:28]: will make these changes in next PR  🗒️

**Pull request merged**